### PR TITLE
[SPARK-34872][SQL] quoteIfNeeded should quote a name which contains non-word characters

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes an issue that `quoteIfNeeded` quotes a name only if it contains `.` or ``` ` ```.
This method should quote it if it contains non-word characters.

### Why are the changes needed?

It's a potential bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New test.

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @cloud-fan @zsxwing

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala [hunk 1]
```diff
@@ -128,4 +128,13 @@ class StringUtilsSuite extends SparkFunSuite with SQLHelper {
       assert(concat.toString === "plan fragment 0plan fragment 1... 15 more characters")
     }
   }
+  test("SPARK-34872: quoteIfNeeded should quote a string which contains non-word characters") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:131]: nit: a blank line before the new test

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 2]
```diff
@@ -148,10 +148,10 @@ package object util extends Logging {
   }
 
   def quoteIfNeeded(part: String): String = {
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:150]: This is a pretty common method. Could you double check if it's okay to change this behavior for other places?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:150]: I think it's only for error message or EXPLAIN result (e.g. `Attribute.sql`). @sarutak can you confirm?

Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:150]: I'm now confirming but, for now,  I don't think this change breaks existing behavior except for display format.

**Comment by sarutak** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala (hunk 2 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:150]: I looked into the following classes which use `quoteIfNeeded`

* ResolveSessionCatalog.apply
* IdentifierHelper.quoted
* MultipartIdentifierHelper.quoted
* DatabaseInSessionCatalog$.unapply
* NamespaceHelper.quoted
* Alias.sql
* AttributeReference.sql
* UnresolvedAttribute.sql
* IdentifierImpl.toString
* IdentifierHelper.quoted

Finally, I think this change doesn't break existing behavior.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by tanelk** [comment]
@sarutak 

I ran git bisect and it seems, that this PR introduced a regression:
```
  test("SPARK-XXXXX: special char in column name") {
    withTempDir { dir =>
      val file = new File(dir, "output.csv")
      val fileContent =
        """
          |a / b
          |val1
          |""".stripMargin

      FileUtils.writeStringToFile(file, fileContent, StandardCharsets.UTF_8)

      spark.read
        .option("header", true)
        .csv(file.toString)
        .where(col("a / b").isNotNull)
        .show()
    }
  }
```
fails with 
```
[info] - SPARK-XXXXX: special char in column name *** FAILED *** (2 seconds, 547 milliseconds)
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 1) (archlinux executor driver): java.lang.IllegalArgumentException: `a / b` does not exist. Available: a / b
[info] 	at org.apache.spark.sql.types.StructType.$anonfun$fieldIndex$1(StructType.scala:306)
[info] 	at scala.collection.immutable.Map$Map1.getOrElse(Map.scala:119)
[info] 	at org.apache.spark.sql.types.StructType.fieldIndex(StructType.scala:305)
[info] 	at org.apache.spark.sql.catalyst.OrderedFilters.$anonfun$predicates$4(OrderedFilters.scala:61)
[info] 	at org.apache.spark.sql.catalyst.OrderedFilters.$anonfun$predicates$4$adapted(OrderedFilters.scala:61)
[info] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
[info] 	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
[info] 	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
[info] 	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
[info] 	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
[info] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
[info] 	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:198)
[info] 	at org.apache.spark.sql.catalyst.OrderedFilters.$anonfun$predicates$3(OrderedFilters.scala:61)
[info] 	at org.apache.spark.sql.catalyst.OrderedFilters.$anonfun$predicates$3$adapted(OrderedFilters.scala:50)
[info] 	at scala.collection.immutable.List.foreach(List.scala:392)
[info] 	at org.apache.spark.sql.catalyst.OrderedFilters.<init>(OrderedFilters.scala:50)
[info] 	at org.apache.spark.sql.catalyst.csv.UnivocityParser.<init>(UnivocityParser.scala:103)
[info] 	at org.apache.spark.sql.execution.datasources.csv.CSVFileFormat.$anonfun$buildReader$1(CSVFileFormat.scala:138)
[info] 	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(FileFormat.scala:148)
[info] 	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(FileFormat.scala:133)
[info] 	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.org$apache$spark$sql$execution$datasources$FileScanRDD$$anon$$readCurrentFile(FileScanRDD.scala:117)
[info] 	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:165)
[info] 	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:94)
[info] 	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
[info] 	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
[info] 	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
[info] 	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:756)
[info] 	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:344)
[info] 	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
[info] 	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
[info] 	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[info] 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[info] 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[info] 	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
[info] 	at org.apache.spark.scheduler.Task.run(Task.scala:131)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:498)
[info] 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1437)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:501)
[info] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info] 	at java.lang.Thread.run(Thread.java:748)
```

**Comment by sarutak** [comment]
@tanelk I think it's a potential bug that filter pushdown cannot handle quoted identifiers.
In fact, if a column contains ``` ` ``` like ```a`b```, the same error happens before this change.
I'll investigate more.

**Comment by cloud-fan** [comment]
fixing in https://github.com/apache/spark/pull/32807

**Comment by sarutak** [comment]
Oh, @cloud-fan is already fixing. Thanks.

**Comment by cloud-fan** [comment]
@sarutak backtick doesn't work because `col(...)` runs a small parser to parse the name, and backtick has a meaning there and you need to escape it.

**Comment by sarutak** [comment]
Ah, of course I mean the case even if we escapes such columns like ```col("`a`b`")```.

**Comment by cloud-fan** [comment]
Ah got it. I've added a test for backtick as well.

**Comment by cloud-fan** [comment]
Hi @mkrasilnikova , can you open a JIRA ticket to report this bug? The Spark codebase is being updated quite frequently and we at least need to know which Spark version has this bug. Thanks!

**Comment by mkrasilnikova** [comment]
Hi @cloud-fan , we started to get this error after we had updated the version from 3.0.2 to 3.2.0. A have just tried with the code from master branch, the bug is still there. 
I opened a ticket https://issues.apache.org/jira/browse/SPARK-37895

**Pull request closed without merging**